File: algorithms/strings/levenshtein-distance/java/LevenshteinDistance.java
public class LevenshteinDistance {

    /**
     * Compute the Levenshtein (edit) distance between two sequences.
     *
     * Input format: [len1, seq1..., len2, seq2...]
     * @param arr input array encoding two sequences
     * @return minimum number of single-element edits
     */
    public static int levenshteinDistance(int[] arr) {
        int idx = 0;
        int len1 = arr[idx++];
        int[] seq1 = new int[len1];
        for (int i = 0; i < len1; i++) seq1[i] = arr[idx++];
        int len2 = arr[idx++];
        int[] seq2 = new int[len2];
        for (int i = 0; i < len2; i++) seq2[i] = arr[idx++];

        int[][] dp = new int[len1 + 1][len2 + 1];

        for (int i = 0; i <= len1; i++) dp[i][0] = i;
        for (int j = 0; j <= len2; j++) dp[0][j] = j;

        for (int i = 1; i <= len1; i++) {
            for (int j = 1; j <= len2; j++) {
                if (seq1[i - 1] == seq2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1];
                } else {
                    dp[i][j] = 1 + Math.min(dp[i - 1][j],
                                   Math.min(dp[i][j - 1], dp[i - 1][j - 1]));
                }
            }
        }

        return dp[len1][len2];
    }

    public static void main(String[] args) {
        System.out.println(levenshteinDistance(new int[]{3, 1, 2, 3, 3, 1, 2, 4})); // 1
        System.out.println(levenshteinDistance(new int[]{2, 5, 6, 2, 5, 6}));       // 0
        System.out.println(levenshteinDistance(new int[]{2, 1, 2, 2, 3, 4}));       // 2
        System.out.println(levenshteinDistance(new int[]{0, 3, 1, 2, 3}));          // 3
    }
}

File: algorithms/strings/levenshtein-distance/scala/LevenshteinDistance.scala
object LevenshteinDistance {

  /**
   * Compute the Levenshtein (edit) distance between two sequences.
   *
   * Input format: [len1, seq1..., len2, seq2...]
   * @param arr input array encoding two sequences
   * @return minimum number of single-element edits
   */
  def levenshteinDistance(arr: Array[Int]): Int = {
    var idx = 0
    val len1 = arr(idx); idx += 1
    val seq1 = arr.slice(idx, idx + len1); idx += len1
    val len2 = arr(idx); idx += 1
    val seq2 = arr.slice(idx, idx + len2)

    val dp = Array.ofDim[Int](len1 + 1, len2 + 1)

    for (i <- 0 to len1) dp(i)(0) = i
    for (j <- 0 to len2) dp(0)(j) = j

    for (i <- 1 to len1) {
      for (j <- 1 to len2) {
        if (seq1(i - 1) == seq2(j - 1)) {
          dp(i)(j) = dp(i - 1)(j - 1)
        } else {
          dp(i)(j) = 1 + math.min(dp(i - 1)(j), math.min(dp(i)(j - 1), dp(i - 1)(j - 1)))
        }
      }
    }

    dp(len1)(len2)
  }

  def main(args: Array[String]): Unit = {
    println(levenshteinDistance(Array(3, 1, 2, 3, 3, 1, 2, 4))) // 1
    println(levenshteinDistance(Array(2, 5, 6, 2, 5, 6)))       // 0
    println(levenshteinDistance(Array(2, 1, 2, 2, 3, 4)))       // 2
    println(levenshteinDistance(Array(0, 3, 1, 2, 3)))          // 3
  }
}
